oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,348 @@
+  dependencies).
+* An automated test suite with reasonably good code coverage, configured to
+  be run on new GitHub pull requests.
+* A service configured to periodically auto-generate a GitHub pull request
+  that tests the output of running ``make upgrade`` (if it results in any
+  changes).  This can either be a service such as `requires.io`_ which tracks
+  new releases of Python package dependencies, or a recurring scheduled job.
+* At least one designated maintainer who receives notifications of the
+  generated pull requests and will merge or fix them as needed.  This
+  maintainer should scan the changelog for each upgraded package to look for
+  changes that merit closer inspection; services like `requires.io`_ and
+  `AllMyChanges.com`_ can make this easier.
+
+.. _requires.io: https://requires.io/
+.. _AllMyChanges.com: https://allmychanges.com/
+
+Installing Dependencies from URLs
+---------------------------------
+
+As noted above, you should generally avoid installing requirements from a URL
+or local directory instead of PyPI.  But there are a few circumstances where
+it can be appropriate:
+
+* You need to test a release candidate of the dependency to make sure it will
+  work with your code.
+* You critically need a fix for a package which has not yet been included in
+  a release, and you cannot arrange for a release to be made in a timely
+  manner.
+
+In most other circumstances, the package should be added to PyPI instead.
+There are several good reasons for this:
+
+* Specified VCS branches, commits, and tags can all be deleted from a
+  repository at any time, suddenly making it impossible to find and install
+  the dependency.
+* Editable requirements (starting with "-e ") are downloaded and/or inspected
+  with each installation of the requirements file, even if the correct version
+  is already installed.  This can significantly slow down updates of installed
+  requirements.
+* Packages installed from local directories don't reflect any changes to
+  package metadata (like required package versions) until the version number
+  is incremented or the package is uninstalled; just installing again doesn't
+  help.
+* Package URLs tend to be long and difficult to read, with the actual name of
+  the package hidden in the middle or not even present at all.
+* As of this writing, ``pip-tools`` still has some bugs in handling packages
+  installed from local directories or URLs that require special care to work
+  around.  `Non-editable URL installations`_ are not supported, and
+  `relative local paths are expanded to absolute paths`_.  These can be
+  partially worked around via a post-processing script for the generated
+  requirements files; an example can be found in `edx-platform`_ at
+  ``scripts/post-pip-compile.sh``.
+
+If you do need to include a package at a URL, it should be editable (start with
+"-e ") and have both the package name and version specified (end with
```

Comment:
The "should be editable" advice is contrary to our current instructions.  You might want to include here, "(because pip-tools doesn't otherwise support URL installations)" or something.

Reply:
Sure, I'll reference the bugs noted above.